### PR TITLE
Disable terminal echo while State Tool is running.

### PR DESCRIPTION
**Pull request by mitchell-as**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-2214" title="DX-2214" target="_blank">TaskDX-2214</a>  As a user I can't interfere with State Tool by pressing keys during read-only actions
  </td></table>
  <br />
 

The exception is for prompts and subshells (which temporarily re-enable echoing).

Note: the survey package does this automatically, so no changes to the prompt package are required.

### CODE REVIEW

File: internal/subshell/bash/bash.go [hunk 1]
```diff
@@ -219,3 +220,17 @@ func (v *SubShell) IsAvailable() bool {
 	}
 	return fileutils.FileExists(rcFile)
 }
+
+func (v *SubShell) TurnOffEcho() {
+	if runtime.GOOS == "windows" {
+		return // not supported
```
Comment (mitchell-as) [inline internal/subshell/bash/bash.go:226]: Windows does not have access to the termios functions needed to disable echo in emulated shells like Git Bash (which is not interactive anyway).

File: internal/subshell/bash/bash.go [hunk 2]
```diff
@@ -219,3 +220,17 @@ func (v *SubShell) IsAvailable() bool {
...
+	}
+	termecho.Off()
+}
+
+func (v *SubShell) TurnOnEcho() {
+	if runtime.GOOS == "windows" {
+		return // not supported
+	}
+	termecho.Off()
```
Comment (MDrakos) [inline internal/subshell/bash/bash.go:235]: Shouldn't these, and the other `TurnOnEcho` functions be calling `termecho.On()`?

File: cmd/state/main.go [hunk 3]
```diff
@@ -207,6 +207,13 @@ func run(args []string, isInteractive bool, cfg *config.Instance, out output.Out
 
 	// Set up conditional, which accesses a lot of primer data
 	sshell := subshell.New(cfg)
+	if isInteractive {
+		// Disable terminal echo while State Tool is running.
+		// Other than in prompts and subshells (which temporarily re-enable echo), user typing should
+		// not interfere with output (e.g. runtime progress bars).
+		sshell.TurnOffEcho()
+		defer sshell.TurnOnEcho()
```
Comment (MDrakos) [inline cmd/state/main.go:215]: Does this also disable ctrl+c and other similar inputs? I don't believe that's something we want to disallow.

Comment (MDrakos) [reply cmd/state/main.go:215]: Also, is it possible that if the state tool errors mid-execution, we could leave the user's shell in a state where they cannot echo?

**Review by MDrakos — CHANGES_REQUESTED** [review]
Something that's not obvious in this PR is how the echo is re-enabled for prompts... how is that handled?

**Comment by mitchell-as** [comment]
As mentioned in the PR description, the survey library automatically handles the flipping on/off of terminal echo. I don't know exactly how it does it, but during testing it did.

File: internal/subshell/bash/bash.go (hunk 2 above)
Comment (mitchell-as) [reply internal/subshell/bash/bash.go]: Oh, how embarrassing. Thanks for catching this.

File: cmd/state/main.go (hunk 3 above)
Comment (mitchell-as) [reply cmd/state/main.go:215]: You didn't see this, but in my original PR (https://github.com/ActiveState/cli/pull/2768), I noted that echo is restored after ^C, so no problem there. The `defer` does its job.

If there is a hard crash, echo will probably be lost, but this is the case for any terminal program that messes with termios, so the only resolution is for the user to exit this terminal and start up a new one.

Comment (MDrakos) [reply cmd/state/main.go:215]: It looks like the `defer` will run even on a panic as long as it's on the function call stack so we should be good even in the event of a crash as this defer is placed ahead of the actual command logic.

**Review by MDrakos — APPROVED** [review]

**Pull request merged**